### PR TITLE
Add ChangeLog repository

### PR DESCRIPTION
**Pull request by kapillamba4**

### CODE REVIEW

File: backend/app/adapter/db/change_log.go [hunk 1]
```diff
@@ -12,10 +12,12 @@ import (
 
 var _ repository.ChangeLog = (*ChangeLogSQL)(nil)
 
+// ChangeLogSQL accesses ChangeLog information in change_log table through SQL.
 type ChangeLogSQL struct {
 	db *sql.DB
 }
 
+// GetChangeByID finds Change from change_log table with given id
```
Comment (byliuyang) [inline backend/app/adapter/db/change_log.go]:
```suggestion
// GetChangeByID finds Change from change_log table with given ID.
```

File: backend/app/adapter/db/migration/13_create_user_change_log_table.sql [hunk 2]
```diff
@@ -0,0 +1,9 @@
+-- +migrate Up
```
Comment (byliuyang) [inline backend/app/adapter/db/migration/13_create_user_change_log_table.sql]: Let's leave this to a different PR with `UserChangeLog` repository.

File: backend/app/adapter/db/migration/12_create_changelog_table.sql [hunk 3]
```diff
@@ -0,0 +1,11 @@
+-- +migrate Up
+CREATE TABLE "change_log"
+(
+    "id"                CHARACTER VARYING(5) PRIMARY KEY,
+    "title"             CHARACTER VARYING (100),
```
Comment (byliuyang) [inline backend/app/adapter/db/migration/12_create_changelog_table.sql]:
```suggestion
    "title"             CHARACTER VARYING(100),
```

File: backend/app/adapter/db/change_log_integration_test.go [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+package db_test
+
+import (
+	"database/sql"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/short-d/short/app/entity"
+
```
Comment (byliuyang) [inline backend/app/adapter/db/change_log_integration_test.go]:
```suggestion
```

File: backend/app/adapter/db/change_log_integration_test.go [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+}
+
+func TestChangeLogSql_GetChangeLog(t *testing.T) {
+	testCases := []struct {
+		name              string
+		tableRows         []changeLogTableRow
+		expectedChangeLog []entity.Change
+	}{
+		{
+			name:              "get complete changelog",
```
Comment (byliuyang) [inline backend/app/adapter/db/change_log_integration_test.go]:
```suggestion
			name:              "get full changelog",
```

File: backend/app/adapter/db/change_log_integration_test.go [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+	}
+}
+
+func TestChangeLogSql_GetChangeLog(t *testing.T) {
+	testCases := []struct {
+		name              string
+		tableRows         []changeLogTableRow
+		expectedChangeLog []entity.Change
+	}{
+		{
```
Comment (byliuyang) [inline backend/app/adapter/db/change_log_integration_test.go]: Let's add an empty change log table case.

File: backend/app/adapter/db/change_log_integration_test.go [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+func TestChangeLogSql_CreateChange(t *testing.T) {
+	testCases := []struct {
+		name                  string
+		tableRows             []changeLogTableRow
+		change                entity.Change
+		expectedChangeLogSize int
+		expectedChange        entity.Change
+	}{
+		{
+			name:                  "Insert a change in changelog",
```
Comment (byliuyang) [inline backend/app/adapter/db/change_log_integration_test.go]:
```suggestion
			name:                  "create a change",
```

File: backend/app/adapter/db/change_log.go [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+	)
+
+	row := c.db.QueryRow(statement, id)
+	var change = entity.Change{}
+
+	err := row.Scan(&change.ID, &change.Title, &change.SummaryMarkdown, &change.ReleasedAt)
+	return change, err
+}
+
+// GetChangeLog retrieves complete changelog from change_log table
```
Comment (byliuyang) [inline backend/app/adapter/db/change_log.go]:
```suggestion
// GetChangeLog retrieves full changelog from change_log table.
```

File: backend/app/adapter/db/change_log.go [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+			return changeLog, err
+		}
+		change.ReleasedAt = utc(change.ReleasedAt)
+		changeLog = append(changeLog, change)
+	}
+
+	return changeLog, nil
+}
+
+// CreateChange inserts a new Change into change_log table
```
Comment (byliuyang) [inline backend/app/adapter/db/change_log.go]:
```suggestion
// CreateChange adds a new Change into change_log table.
```

File: backend/app/adapter/db/change_log.go [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+
+	if err != nil {
+		return entity.Change{}, err
+	}
+
+	return c.GetChangeByID(newChange.ID)
+}
+
+// NewChangeLogSQL creates ChangeLogSQL
+func NewChangeLogSQL(db *sql.DB) *ChangeLogSQL {
```
Comment (byliuyang) [inline backend/app/adapter/db/change_log.go]: Can this be `*ChangeLogSQL` ( value ) instead of `*ChangeLogSQL` ( pointer )?

File: backend/app/adapter/db/migration/12_create_changelog_table.sql [hunk 11]
```diff
@@ -0,0 +1,11 @@
+-- +migrate Up
+CREATE TABLE "change_log"
+(
+    "id"                CHARACTER VARYING(5) PRIMARY KEY,
```
Comment (alldroll) [inline backend/app/adapter/db/migration/12_create_changelog_table.sql]: Why do we use `VARCHAR` instead of `SERIAL` for `id`?

File: backend/app/adapter/db/change_log.go [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+
+	row := c.db.QueryRow(statement, id)
+	var change = entity.Change{}
+
+	err := row.Scan(&change.ID, &change.Title, &change.SummaryMarkdown, &change.ReleasedAt)
+	return change, err
+}
+
+// GetChangeLog retrieves full changelog from change_log table.
+func (c ChangeLogSQL) GetChangeLog() ([]entity.Change, error) {
```
Comment (alldroll) [inline backend/app/adapter/db/change_log.go]: Are you sure it is safe? I mean, `change_log` table is assumed to be `huge` in order to hold all changes. 
And I don't when we need to get the whole dataset. Maybe we should add pagination or other search criteria?

File: backend/app/adapter/db/change_log.go [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+		newChange.Title,
+		newChange.SummaryMarkdown,
+		newChange.ReleasedAt,
+	)
+
+	if err != nil {
+		return entity.Change{}, err
+	}
+
+	return c.GetChangeByID(newChange.ID)
```
Comment (alldroll) [inline backend/app/adapter/db/change_log.go]: What is the purpose to make this query if we already saved the entity and know exactly how it should look?

File: backend/app/adapter/db/change_log.go [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+
+	_, err := c.db.Exec(
+		statement,
+		newChange.ID,
+		newChange.Title,
+		newChange.SummaryMarkdown,
+		newChange.ReleasedAt,
+	)
+
+	if err != nil {
```
Comment (alldroll) [inline backend/app/adapter/db/change_log.go]: Maybe we should add a handler for the case when `ID` is already busy?

File: backend/app/adapter/db/change_log.go (hunk 13 above)
Comment (kapillamba4) [reply backend/app/adapter/db/change_log.go]: I just wanted to be sure that we inserted the change in change_log table in case db.Exec fails without any error... i don't think it will happen though

File: backend/app/adapter/db/change_log.go (hunk 12 above)
Comment (kapillamba4) [reply backend/app/adapter/db/change_log.go]: We had discussion on issue here: https://github.com/short-d/short/issues/566 and a new issue was created: https://github.com/short-d/short/issues/595, pagination will be added with its corresponding PR

File: backend/app/adapter/db/change_log.go (hunk 14 above)
Comment (kapillamba4) [reply backend/app/adapter/db/change_log.go]: What does it mean by `ID` is busy?

File: backend/app/adapter/db/migration/12_create_changelog_table.sql (hunk 11 above)
Comment (kapillamba4) [reply backend/app/adapter/db/migration/12_create_changelog_table.sql]: We are generating Key manually using KeyGenerator

File: backend/app/adapter/db/change_log.go (hunk 14 above)
Comment (alldroll) [reply backend/app/adapter/db/change_log.go]: A row with the given ID is already in database

File: backend/app/adapter/db/change_log.go (hunk 13 above)
Comment (alldroll) [reply backend/app/adapter/db/change_log.go]: I think it is a redundant check. It is an issue of DB driver, not ours.

File: backend/app/adapter/db/migration/12_create_changelog_table.sql (hunk 11 above)
Comment (alldroll) [reply backend/app/adapter/db/migration/12_create_changelog_table.sql]: Oh, got it. Cool:)

File: backend/app/adapter/db/change_log.go (hunk 13 above)
Comment (kapillamba4) [reply backend/app/adapter/db/change_log.go]: @byliuyang what's your opinion, shall we remove GetChangeByID method?

File: backend/app/adapter/db/change_log_integration_test.go [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+		{
+			name:           "ID does not exist",
+			id:             "12345",
+			tableRows:      []changeLogTableRow{{id: "12346", title: "title 2", summaryMarkdown: "summary 2"}},
+			expectedChange: entity.Change{},
+		},
+		{
+			name:           "ID exists",
+			id:             "12345",
+			tableRows:      []changeLogTableRow{{id: "12346", title: "title 2", summaryMarkdown: "summary 2"}, {id: "12345", title: "title 1", summaryMarkdown: "summary 1"}},
```
Comment (byliuyang) [inline backend/app/adapter/db/change_log_integration_test.go]:
```suggestion
			tableRows:      []changeLogTableRow{
			{id: "12346", title: "title 2", summaryMarkdown: "summary 2"}, 
			{id: "12345", title: "title 1", summaryMarkdown: "summary 1"},
		},
```

File: backend/app/adapter/db/change_log_integration_test.go [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+func TestChangeLogSql_GetChangeLog(t *testing.T) {
+	testCases := []struct {
+		name              string
+		tableRows         []changeLogTableRow
+		expectedChangeLog []entity.Change
+	}{
+		{
+			name:              "get full changelog",
+			tableRows:         []changeLogTableRow{{id: "12346", title: "title 2", summaryMarkdown: "summary 2"}, {id: "12345", title: "title 1", summaryMarkdown: "summary 1"}},
+			expectedChangeLog: []entity.Change{{ID: "12346", Title: "title 2", SummaryMarkdown: "summary 2"}, {ID: "12345", Title: "title 1", SummaryMarkdown: "summary 1"}},
```
Comment (byliuyang) [inline backend/app/adapter/db/change_log_integration_test.go]:
```suggestion
			tableRows:         []changeLogTableRow{{id: "12346", title: "title 2", summaryMarkdown: "summary 2"}, {id: "12345", title: "title 1", summaryMarkdown: "summary 1"}},
			expectedChangeLog: []entity.Change{
			{ID: "12346", Title: "title 2", SummaryMarkdown: "summary 2"}, 
			{ID: "12345", Title: "title 1", SummaryMarkdown: "summary 1"},
			},
```

File: backend/app/adapter/db/change_log_integration_test.go [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+	testCases := []struct {
+		name                  string
+		tableRows             []changeLogTableRow
+		change                entity.Change
+		expectedChangeLogSize int
+		expectedChange        entity.Change
+	}{
+		{
+			name:                  "create a change",
+			tableRows:             []changeLogTableRow{{id: "12345", title: "title 1", summaryMarkdown: "summary 1"}, {id: "12346", title: "title 2", summaryMarkdown: "summary 2"}},
```
Comment (byliuyang) [inline backend/app/adapter/db/change_log_integration_test.go]:
```suggestion
			tableRows:             []changeLogTableRow{
			{id: "12345", title: "title 1", summaryMarkdown: "summary 1"}, 
			{id: "12346", title: "title 2", summaryMarkdown: "summary 2"},
			},
```

File: backend/app/adapter/db/migration/12_create_changelog_table.sql (hunk 11 above)
Comment (byliuyang) [reply backend/app/adapter/db/migration/12_create_changelog_table.sql]: This is for scalability reason.

File: backend/app/adapter/db/change_log.go (hunk 10 above)
Comment (kapillamba4) [reply backend/app/adapter/db/change_log.go]: Everywhere else in adapter/db we are using pointer

Comment (byliuyang) [reply backend/app/adapter/db/change_log.go]: The URL SQL repo uses pointer because of mutation operation here: 
```go
func (u *URLSql) Create(url entity.URL) error {
```

File: backend/app/adapter/db/change_log.go (hunk 13 above)
Comment (byliuyang) [reply backend/app/adapter/db/change_log.go]: Please replace this one with the struct we passed in as @alldroll suggested for now.

Comment (kapillamba4) [reply backend/app/adapter/db/change_log.go]: Will do that now

File: backend/app/adapter/db/change_log_integration_test.go [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+	"testing"
+	"time"
+
+	"github.com/short-d/app/mdtest"
+	"github.com/short-d/short/app/adapter/db"
+	"github.com/short-d/short/app/adapter/db/table"
+	"github.com/short-d/short/app/entity"
+)
+
+var insertChangeLogRowSQL = fmt.Sprintf(`INSERT INTO %s (%s, %s, %s, %s) VALUES ($1, $2, $3, $4)`,
```
Comment (byliuyang) [inline backend/app/adapter/db/change_log_integration_test.go]:
```suggestion
var insertChangeLogRowSQL = fmt.Sprintf(`
INSERT INTO %s (%s, %s, %s, %s) 
VALUES ($1, $2, $3, $4)`,
```
